src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java

```diff
@@ -0,0 +1,121 @@
+package de.tum.in.www1.artemis.service.hestia;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.ProgrammingExercise;
+import de.tum.in.www1.artemis.domain.hestia.ProgrammingExerciseTask;
+import de.tum.in.www1.artemis.repository.ProgrammingExerciseTestCaseRepository;
+import de.tum.in.www1.artemis.repository.hestia.CodeHintRepository;
+import de.tum.in.www1.artemis.repository.hestia.ProgrammingExerciseTaskRepository;
+
+@Service
+public class ProgrammingExerciseTaskService {
+
+    private ProgrammingExerciseTaskRepository programmingExerciseTaskRepository;
+
+    private ProgrammingExerciseTestCaseRepository programmingExerciseTestCaseRepository;
+
+    private CodeHintRepository codeHintRepository;
+
+    /**
+     * Pattern that is used to extract the tasks (capturing group "name") and test case names (capturing group "tests") from the problem statement.
+     * Example: "[task][Implement BubbleSort](testBubbleSort,testBubbleSortHidden)". Following values are extracted by the named capturing groups:
+     * - name: "Implement BubbleSort"
+     * - tests: "testBubbleSort,testBubbleSortHidden"
+     */
+    private final Pattern taskPatternForProblemStatementMarkdown = Pattern.compile("\\[task]\\[(?<name>[^\\[\\]]+)]\\((?<tests>.*)\\)");
+
+    public ProgrammingExerciseTaskService(ProgrammingExerciseTaskRepository programmingExerciseTaskRepository,
+            ProgrammingExerciseTestCaseRepository programmingExerciseTestCaseRepository, CodeHintRepository codeHintRepository) {
+        this.programmingExerciseTaskRepository = programmingExerciseTaskRepository;
+        this.programmingExerciseTestCaseRepository = programmingExerciseTestCaseRepository;
+        this.codeHintRepository = codeHintRepository;
+    }
+
+    /**
+     * Deletes a ProgrammingExerciseTask together with its CodeHints
+     * This has to be manually done, as there is no orphanRemoval between the two entities
+     *
+     * @param task The task to delete
+     */
+    public void delete(ProgrammingExerciseTask task) {
+        var codeHints = codeHintRepository.findByTaskId(task.getId());
+        codeHintRepository.deleteAll(codeHints);
+        programmingExerciseTaskRepository.delete(task);
+    }
+
+    /**
+     * Extracts all tasks from the problem statement of an exercise and saves them to the database.
+     * All tasks that no longer exist will be deleted.
+     * If there is already a task with the same test cases as a new one, but with a different name the existing one will be renamed.
+     *
+     * @param exercise The programming exercise to extract the tasks from
+     */
+    public void updateTasksFromProblemStatement(ProgrammingExercise exercise) {
+        var tasksToBeDeleted = programmingExerciseTaskRepository.findByExerciseIdWithTestCases(exercise.getId());
```

Comment:
I just add this as a new comment as I think the other comment got lost.
> Thanks for adapting the names but I don't think you quite understood my idea:
Now in line 60 you define the variable tasksToBeDeleted but actually, these are the existingTasks.
I would ask you to not just override the meaning of your variables but just create new sets from the ones you create in lines 60-61.
This makes it easier to read certain parts of the code because otherwise, you'd have to read the whole code just to understand that the meaning of a variable changes despite having the same name.